confluent-kafka-plugins/src/main/java/io/cdap/plugin/confluent/streaming/source/ConfluentStreamingSource.java

```diff
@@ -79,7 +102,32 @@ public JavaDStream<StructuredRecord> getStream(StreamingContext context) throws
     collector.getOrThrowException();
 
     context.registerLineage(conf.referenceName);
-    return ConfluentStreamingSourceUtil.getStructuredRecordJavaDStream(context, conf, outputSchema, collector);
+    JavaInputDStream<ConsumerRecord<Object, Object>> javaInputDStream = ConfluentStreamingSourceUtil
+      .getConsumerRecordJavaDStream(context, conf, outputSchema, collector, getStateSupplier(context));
+
+    JavaDStream<StructuredRecord> javaDStream;
+    javaDStream = ConfluentStreamingSourceUtil.getStructuredRecordJavaDStream(javaInputDStream,
+      new ConfluentStreamingSourceUtil.RecordTransform(conf, outputSchema));
+
+    if (conf.getSchemaRegistryUrl() != null) {
+      ConfluentStreamingSourceUtil.AvroRecordTransform transform =
+        new ConfluentStreamingSourceUtil.AvroRecordTransform(conf, outputSchema);
+      javaDStream = ConfluentStreamingSourceUtil.getStructuredRecordJavaDStream(javaInputDStream, transform);
+    }
+
+    if (!context.isStateStoreEnabled()) {
+      // Return the serializable Dstream in case checkpointing is enabled.
+      return javaDStream;
+    }
+
+    // Use the DStream that is state aware
+
+    ConfluentDStream confluentDStream = new ConfluentDStream(context.getSparkStreamingContext().ssc(),
+                                                             javaInputDStream.inputDStream(),
+                                                             ConfluentStreamingSourceUtil
+                                                               .getRecordTransformFunction(conf, outputSchema),
```

Comment:
Shouldn't we check for schema registry URL here? Use a variable when checking first and use the same for both with state and without state.

Reply:
The schema registry URL check has been done here:

```
if (conf.getSchemaRegistryUrl() != null) {
      ConfluentStreamingSourceUtil.AvroRecordTransform transform =
        new ConfluentStreamingSourceUtil.AvroRecordTransform(conf, outputSchema);
      javaDStream = ConfluentStreamingSourceUtil.getStructuredRecordJavaDStream(javaInputDStream, transform);
}
```
**Do we need to again check for it while creating the ConfluentDStream?**
ConfluentDStream takes InputDStream as an argument whereas, during the schema registry URL check, the transformation returns JavaDStream.

Reply:
Moved the following block inside `if (!context.isStateStoreEnabled())` check. Additionally, removed the `javaDStream` and `transform` variables. b08abc3dce38d301121324f6ed52e1e76e8c75c6

```
if (conf.getSchemaRegistryUrl() != null) {
        return javaInputDStream.transform(new ConfluentStreamingSourceUtil.AvroRecordTransform(conf, outputSchema));
} else {
        return javaInputDStream.transform(new ConfluentStreamingSourceUtil.RecordTransform(conf, outputSchema));
}
```

Also, added the `conf.getSchemaRegistryUrl()` check to the `ConfluentStreamingSourceUtil.getRecordTransformFunction(conf, outputSchema)` method

```
if (conf.getSchemaRegistryUrl() != null) {
      return new AvroFunction(conf, outputSchema);
}
```